### PR TITLE
FRC-0046 First draft of fungible token standard.

### PR DESCRIPTION
**Pull request by anorth**
This is the first draft of a fungible token standard API for native actors, as discussed in https://github.com/filecoin-project/FIPs/discussions/407

An implementation of this standard is in development at https://github.com/helix-onchain/filecoin/tree/main/fil_fungible_token

FYI @jsuresh, @alexytsu @ZenGround0

### CODE REVIEW

File: FRCs/frc-tokenstandard.md [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+// Returns the balance of an address.
+// Must be non-negative.
+fn balance_of(owner: Address) -> TokenAmount
+
+// Transfers tokens from caller to another address.
+// Amount must be non-negative (but can be zero).
+// Transferring to the caller must be treated as a normal transfer.
+// Returns the sender's remaining balance.
+// Aborts if the receiver hook on the `to` address aborts.
+fn transfer({to: Address, amount: TokenAmount, data: Bytes}) -> TokenAmount
```
Comment (hunjixin) [inline FRCs/frc-tokenstandard.md:86]: what is the usage of data parameter?

Comment (hunjixin) [reply FRCs/frc-tokenstandard.md:86]: any piece of data for the receiver hook?

File: FRCs/frc-tokenstandard.md [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+
+Methods are to be presented according to a calling convention such as that discussed in [#382](https://github.com/filecoin-project/FIPs/discussions/382), 
+so this recommendation depends on that.
+
+### Interface FRC-XXX
+
+```
+// Token amounts serialize as var-int, and have an implied 18 decimals precision,
+// matching the native token.
+type TokenAmount = BigInt
```
Comment (Kubuxu) [inline FRCs/frc-tokenstandard.md:61]: There is an argument for using uint128 (or two uint64) for the token amount. I think FVM evaluated both at some point.

File: FRCs/frc-tokenstandard.md (hunk 1 above)
Comment (Kubuxu) [reply FRCs/frc-tokenstandard.md:86]: Yes, you can imagine a use-case where you pay for something with a token and you want to attach tracking information to it.

**Comment by Kubuxu** [comment]
I see that `decimals()` was discarded. Should we provide something like `granularity()` from EIP-777. While rejecting transactions internally is the correct thing to do, I can see a need for knowing what is the granularity of a given token. For example a generic wallet application and users wanting to transfer half of their tokens.

File: FRCs/frc-tokenstandard.md (hunk 1 above)
Comment (anorth) [reply FRCs/frc-tokenstandard.md:86]: Yes, thanks I need to clarify that in the text.

**Comment by jsuresh** [comment]
@Kubuxu - I understood the purpose of granularity differently. Love a few more examples from you as I may have missed important use cases.

my understanding is everything has 18 DP as  this being different is an endless source of client bugs. However, in certain use cases it doesn’t make sense to hold or transact partial tokens.

my goto example is games. Eg. Rupees in Zelda have to be ‘whole’ for use in game and therefore trading.

My assumption was in filecoin this kind of use case would be rare. However, would love to hear other motivations/examples of ways you think this would come in handy.

**Comment by anorth** [comment]
Let's take granularity discussion to https://github.com/filecoin-project/FIPs/discussions/407#discussioncomment-3418757 and leave this review for FRC style/completeness.

**Comment by jennijuju** [comment]
hi - this is our FRC-0046

**Comment by anorth** [comment]
I have pushed a new commit which:
- adds granularity method
- expands the return values from some methods
- adds tokenData separate from operatorData for the receiver hook
- clarifies a few things

**Comment by jennijuju** [comment]
is this ready for review? for fip editor?

**Comment by anorth** [comment]
@jennijuju I think it is, but I think there's one more round of minor tweaks coming, so let's wait just a week more.

**Comment by anorth** [comment]
I have pushed updates to
- make the receiver hook universal
- use FRC-42 compliant method names

Let's leave this for a day or two, but then I think it's ready for merge. Of course it's still in draft.

**Pull request merged**